### PR TITLE
Ignore Kafka tombstones by default

### PR DESCRIPTION
**Pull request by rvesse**
Improves `KafkaEventSource` behaviour so tombstone events (those with null values) which are used to apply "deletes" to a Kafka topic are ignored unless configured to not ignore them.  This allows an application to silently move past tombstone events without needing any application visible code changes.

This was some experimental work motivated by @danieldaviestelicent's recent investigation into Kafka crypto-shredding/tombstoning to understand what our current behaviour around tombstones was and to modify it to align with the vision of the Data Management work where deletions from Kafka (via tombstone events) should effectively be invisible to actual applications built atop these libraries.

### CODE REVIEW

File: event-sources/event-sources-core/src/main/java/io/telicent/smart/cache/sources/memory/SimpleEvent.java [hunk 1]
```diff
@@ -39,6 +39,7 @@ public class SimpleEvent<TKey, TValue> implements Event<TKey, TValue> {
     private final TKey key;
     private final TValue value;
     @ToString.Exclude
+    @SuppressWarnings("rawtypes")
     private final EventSource source;
```
Comment (robwtelicent) [inline event-sources/event-sources-core/src/main/java/io/telicent/smart/cache/sources/memory/SimpleEvent.java:43]:
```suggestion
private final EventSource<?,?> source;
```
I think this is preferable as we are explicitly saying we don't know the type - rather than just ignoring it.

**Review by robwtelicent — APPROVED** [review]

File: event-sources/event-sources-core/src/main/java/io/telicent/smart/cache/sources/memory/SimpleEvent.java (hunk 1 above)
Comment (rvesse) [reply event-sources/event-sources-core/src/main/java/io/telicent/smart/cache/sources/memory/SimpleEvent.java:43]: Good point, it's not as trivial as just this change as there's a paired method `EventSource.processed()` that also needs to change but it makes the code more type safe and reduces compiler warnings so have gone ahead and applied that change in my latest commit

**Review by TelicentPaul — CHANGES_REQUESTED** [review]
I'd like to see better test coverage - say, a tombstone only batch or a batch that ends with a tombstone, as I'm not sure the logic behaves as it should.  I don't think we will commit the offsets properly. 

As a follow-on, the polling might miss events in those same situations. 

They're admittedly unlikely to ever crop up but could cause serious problems if they did. I doubt we'd be able to diagnose it at the time too. 

I'm committing some hacked together tests to illustrate better

**Comment by rvesse** [comment]
> I'd like to see better test coverage - say, a tombstone only batch or a batch that ends with a tombstone, as I'm not sure the logic behaves as it should. I don't think we will commit the offsets properly.
> 
> As a follow-on, the polling might miss events in those same situations.
> 
> They're admittedly unlikely to ever crop up but could cause serious problems if they did. I doubt we'd be able to diagnose it at the time too.
> 
> I'm committing some hacked together tests to illustrate better

Will take a look and refine the approach accordingly

**Comment by TelicentPaul** [comment]
> > I'd like to see better test coverage - say, a tombstone only batch or a batch that ends with a tombstone, as I'm not sure the logic behaves as it should. I don't think we will commit the offsets properly.
> > As a follow-on, the polling might miss events in those same situations.
> > They're admittedly unlikely to ever crop up but could cause serious problems if they did. I doubt we'd be able to diagnose it at the time too.
> > I'm committing some hacked together tests to illustrate better
> 
> Will take a look and refine the approach accordingly

Apologies. Not what you want before going on holiday = )

**Comment by rvesse** [comment]
@TelicentPaul So I have made some of the changes **but** I think one of your test scenarios of ignoring tombstones with manual commits and a trailing tombstone is an edge case that applications simply have to choose whether they care about.  If they are ignoring tombstones then they may have to reprocess (ignore again) those tombstones upon restart but provided there aren't crazy numbers of them at the end of a topic this shouldn't be an issue IMO.

I fleshed out the tests with some other scenarios, in the auto-commit case the automatically committed offsets do include the ignored tombstones.  So it's only the manual commit case that could be problematic and I think that's primarily an application concern so I've explicitly documented that.

The `poll()` behaviour was non-intuitive as your test illustrated, generally `poll()` is getting called in a loop so if we get a fully tombstone batch we'd tend to pick up normal events on a subsequent call.  However, if there's lots of tombstones this would be less than helpful behaviour.  I have modified this so that if we get and discard a full batch of events we'll immediately try and `poll()` Kafka again **provided** we have timeout remaining.  If all we get is tombstones by the time the supplied timeout has elapsed then we'll ultimately return `null` from the `poll()` call.

**Review by TelicentPaul — APPROVED** [review]
Looks good

**Review by robwtelicent — APPROVED** [review]

**Comment by rvesse** [comment]
A heap space error is reported by the CI/CD pipeline suggesting that the latest change may run afoul of some edge cases that cause us to exhaust heap, investigating

**Review by robwtelicent — APPROVED** [review]

**Review by robwtelicent — APPROVED** [review]
